File: Neetcode/LargestRectHist.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        index = []
        height = []
        area = 0
        
        for i, h in enumerate(heights):
            if not height:
                height.append(h)
                index.append(i)
            else:
                if h >= height[-1]:
                    height.append(h)
                    index.append(i)
                else:
                    while height and h < height[-1]:
                        area = max(area, height[-1]*(i-index[-1]))
                        height.pop()
                        last_in = index.pop()
                    height.append(h)
                    index.append(last_in)
                    # height[-1] = h
            
            # print(i, h, index, height, area)

        # print(index, height, area)

        while height:
            area = max(area, height[-1]*(len(heights) - index[-1]))
            height.pop()
            index.pop()
            # print(index, height, area)

        # print(index, height, area)

        return area

File: Neetcode/LargestRectHist.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int, int>> blocks;
        int area(0);
        for (int i=0; i<heights.size(); i++){
            if (blocks.empty()){
                blocks.push(make_pair(heights[i], i));
            }
            else{
                if (blocks.top().first <= heights[i]){
                    blocks.push(make_pair(heights[i], i));
                }
                else{
                    int last_index(-1);
                    while (!blocks.empty() && blocks.top().first > heights[i]){
                        area = max(area, blocks.top().first * (i-blocks.top().second));
                        last_index = blocks.top().second;
                        blocks.pop();
                    }
                    blocks.push(make_pair(heights[i], last_index));
                }
            }
        }
        while (!blocks.empty()){
            area = max(area, (int)(blocks.top().first * (heights.size()-blocks.top().second)));
            blocks.pop();
        }
        return area;
    }
};
